### DISCUSSION TITLE
Is there a way to achieve a bidirectional link similar to that between two actors in Erlang

### DISCUSSION THREAD

**Discussion by phil-skillwon** [slawlor/ractor#166] [Q&A] [answered]
Two actors in ractor are linked using the link method, which is unidirectional. Is there a way to achieve a bidirectional link similar to that between two actors in Erlang, where no matter who exits first, the other actor will receive an exit event.

**Comment by slawlor** [comment, marked as answer]
You can just call link in reverse to bi directionally connect them. I suppose the current solution is more akin to `monitor` in Erlang.

But you could do

```rust
a.link(b);
b.link(a);
```

**Reply by phil-skillwon** [reply]
Before starting this discussion, I had already conducted tests. Here is how I tested it:

sampler.get_cell().link(proc.get_cell());
proc.get_cell().link(sampler.get_cell());
concurrency::sleep(concurrency::Duration::from_secs(3)).await;
// sampler.stop(Some("user shutdown".to_owned()));
proc.stop(Some("user shutdown".to_owned()));

However, during runtime, it crashed with the following error message: 
"thread 'tokio-runtime-worker' has overflowed its stack, fatal runtime error: stack overflow". 
Did I use the link method incorrectly?

**Reply by slawlor** [reply]
Hmm I think I have a suspicion what's going on here, and this is a legitimate bug. Do you mind opening an issue to this effect and I'll investigate? Perhaps I need to add some other support for bidirectional linking

**Reply by phil-skillwon** [reply]
After updating to version 0.9.1, everything is working fine. 
However, I still have a question: 
a.link(b); 
b.link(a); 
After establishing this bidirectional link, when a exits, the handle_supervisor_evt method of b is not being called. The same is true in reverse: when b exits, the handle_supervisor_evt method of a is also not being called.

**Reply by slawlor** [reply]
That's because they're not just monitoring each other, they're supervisors of each other. So if one supervisor dies it kills its children, before it can handle any other work, like handling the supervision event. Our supervision model isn't exactly implemented like Erlang's

For now that's a limitation I think until we work out a "monitoring" flow. Which say would notify without killing. I have some changes in a draft branch but it's not fully fleshed out yet

**Reply by nsengupta** [reply]
Not sure if my Q falls in this category, but behaviour-wise it is similar.

Is there a Parent-Child Actor specific communication idiom in _ractor_? For example, can an Actor pass a message to its Supervisor (its creator or **Parent** actor) through any specific method?

Because of my past experience with Akka, where any actor can send an asynchronous (of course) message to its parent by saying `parent ! some_message` ( Scala syntax ), I am considering this in _ractor_ too.  In Akka, `parent` is automatically available to any Actor.

Sorry if this is something I have missed in the documentation. In that case, please direct me towards it.

**Reply by slawlor** [reply]
No there's no default parent threaded into the child. You can pass the `ActorRef<_>` yourself down to the child (which is commonly done), but it's not done automatically.
